### PR TITLE
Remove-AzVmDataDisk  -ForceDetach support

### PR DESCRIPTION
**Pull request by grizzlytheodore**
## Description

Added parameter `-ForceDetach` to `Remove-AzVMDataDisk` cmdlet.

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Review by dolauli — APPROVED** [review]

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1 [hunk 1]
```diff
@@ -7538,10 +7538,25 @@ function Test-AddVMDataDisk
...
     # Validate
     Assert-AreEqual $vmConfig.StorageProfile.DataDisks[0].SourceResource.id "testSourceResourceId"
+
+    # test Remove-AzVMDataDisk
+    $vmConfig = Remove-AzVMDataDisk -VM $vmConfig -DataDiskNames @('datadisk1') -ForceDetach
+
+    # Validate
+    $dataDisks = $vmConfig.StorageProfile.DataDisks
+    Assert-NotNullOrEmpty $dataDisks
+    Assert-AreEqual $dataDisks.Count 3
```
Comment (haagha) [inline src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1:7554]: curious if 3 disks were added and 1 was removed, why is the count still 3?

File: src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs [hunk 2]
```diff
@@ -53,20 +60,36 @@ public override void ExecuteCmdlet()
                     var disks = storageProfile.DataDisks.ToList();
                     var comp = StringComparison.OrdinalIgnoreCase;
 
-                    if (DataDiskNames == null)
-                    {
-                        disks.Clear();
+                    if (ForceDetach != true) {
```
Comment (haagha) [inline src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs:63]: it's probably a better practice for code readability to reference parameters explicitly by **this**.ForceDetach. so not to be confused by any local variables inside functions.

File: src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs [hunk 3]
```diff
@@ -53,20 +60,36 @@ public override void ExecuteCmdlet()
...
+                            foreach (var diskName in DataDiskNames){
+                                disks.RemoveAll(d => string.Equals(d.Name, diskName, comp));
+                            }
+                        }
                     }
-                    else
-                    {
-                        foreach (var diskName in DataDiskNames)
+                    else{
+                        if (DataDiskNames == null){
```
Comment (haagha) [inline src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs:74]: should we also add check for empty string/empty List?

**Review by haagha — COMMENTED** [review]
A few comments.

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1 [hunk 4]
```diff
@@ -7538,10 +7538,25 @@ function Test-AddVMDataDisk
...
+    Assert-AreEqual $dataDisks[1].ToBeDetached $true
+    Assert-AreEqual $dataDisks[1].DetachOption "ForceDetach"
+    Assert-AreNotEqual $dataDisks[2].ToBeDetached $true
+    Assert-AreNotEqual $dataDisks[2].DetachOption "ForceDetach"
+
```
Comment (haagha) [inline src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1:7559]: I would also like to test: Remove-AzVMDataDisk -VM $vmConfig -ForceDetach

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1 (hunk 1 above)
Comment (grizzlytheodore) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1:7554]: so for force detach, I have to pass in "detachOption = forceDetach" property of the data disk. 
So I cant remove the data disk object, but it gets detached by setting "toBeDetached" property value to be true

File: src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs (hunk 3 above)
Comment (grizzlytheodore) [reply src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs:74]: this is only for the case of empty list, string as this is a mandatory parameter

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1 (hunk 1 above)
Comment (grizzlytheodore) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1:7554]: so Count is still 3, but one of them have toBeDetached == true and will be removed

File: src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1 (hunk 4 above)
Comment (grizzlytheodore) [reply src/Compute/Compute.Test/ScenarioTests/VirtualMachineTests.ps1:7561]: updated to include that case

**Review by haagha — APPROVED** [review]

**Review by dolauli — APPROVED** [review]

**Pull request merged**